lib/Terminal/Print.pm6

```diff
@@ -99,9 +94,39 @@ class Terminal::Print {
         }
     }
 
-    method root-widget() {
+    method setup(:$reset = False) {
+        $!move-cursor  //= move-cursor-template($!cursor-profile);
+        if $reset {
+            $!columns = columns();
+            $!rows = rows();
+            $!current-grid = Terminal::Print::Grid.new( $!columns, $!rows, :$!move-cursor );
+        }
+        else {
+            $!columns //= columns();
+            $!rows //= rows();
+            $!current-grid //= Terminal::Print::Grid.new( $!columns, $!rows, :$!move-cursor );
+        }
+        if %!grid-name-map<.default>.defined {
+            my $gidx = %!grid-name-map<.default>;
+            my $old-grid = @!grids[$gidx];
+            unless $!current-grid<> =:= $old-grid {
+                my $root-widget = %!root-widget-map{$old-grid}:delete;
+                %!root-widget-map{$!current-grid} = $root-widget;
+                $root-widget.replace-grid: @!grids[$gidx] = $!current-grid;
+            }
+        }
+        else {
+            self.add-grid: '.default', new-grid => $!current-grid;
+        }
+    }
+
+    multi method root-widget(Terminal::Print::Widget $widget-type?) {
         my $grid = self.current-grid;
-        %!root-widget-map{$grid} ||= Terminal::Print::Widget.new-from-grid($grid);
+        %!root-widget-map{$grid} ||= $widget-type.WHAT.new-from-grid($grid);
```

Comment:
Why not make constrain the parameter to :U and avoid the .WHAT indirection?

Reply:
Left-over from earlier implementation before I made it a multi. Will fix.